File: code/ch20/20.2.5. 127.ladderLength.cpp
#include <vector>
#include <string>
#include <map>
#include <queue>
#include <set>
using namespace std;

class Solution {
public:
  int ladderLength(string beginWord, string endWord,
      vector<string>& wordList) {
    queue<string> q;
    q.push(beginWord);
    set<string> visited;
    int steps = 1, n = wordList.size(), L = beginWord.length();
    map<string, vector<string> > wizards;
    for (int i = 0; i < n; i++) {
      string word = wordList[i];
      for (int j = 0; j < L; j++) {
        string key = word.substr(0, j) + string("*") + \
                     word.substr(j + 1,
                         word.length() - j - 1);
        if (!wizards.count(key))
          wizards[key] = vector<string>();
        wizards[key].push_back(word);
      }
    }
    while (!q.empty()) {
      int last_q_size = q.size();
      for (int flg = 0; flg < last_q_size; flg++) {
        string cur = q.front();
        q.pop();
        if (visited.count(cur)) continue;
        visited.insert(cur);
        if (cur == endWord) return steps;
        for (int i = 0; i < L; i++) {
          string key = cur.substr(0, i) + string("*") + \
                       cur.substr(i + 1,
                           cur.length() - i - 1);
          if (!wizards.count(key)) continue;
          for (int j = 0; j < wizards[key].size(); j++)
            q.push(wizards[key][j]);
        }
      }
      steps += 1;
    }
    return 0;
  }
};

File: code/ch20/20.2.5. 127.ladderLength.py
from collections import defaultdict


class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        queue = [beginWord]
        visited = set()
        steps = 1
        n = len(wordList)
        L = len(beginWord)
        wizards = defaultdict(list)
        for i in range(n):
            word = wordList[i]
            for j in range(L):
                wizards[word[:j] + "*" + word[j + 1 :]].append(word)

        while len(queue) > 0:
            for _ in range(len(queue)):
                cur = queue.pop(0)
                if cur in visited:
                    continue
                visited.add(cur)
                if cur == endWord:
                    return steps
                for i in range(L):
                    for word in wizards.get(cur[:i] + "*" + cur[i + 1 :], []):
                        queue.append(word)

            steps += 1
        return 0
